File: src/hyperliquid_python/services/close_order.py
from models.webhook import WebhookPayload
from models.common import OrderResult
from helpers.error_handler import AppError
from constants.http import HTTP
from helpers.telegram import send_notification
from helpers.hyperliquid_helpers import (
    get_env_config, create_clients, get_asset_info, 
    get_strategy_position, get_order_pnl, extract_order_id
)
from helpers.market_price_helpers import get_formatted_market_price
from repositories.dynamo_repository import update_order

def close_order(signal: WebhookPayload) -> OrderResult:
    try:
        print(f"[close_order] Received signal: {signal.model_dump()}")

        private_key, user_address, is_testnet = get_env_config()
        exchange_client, info_client = create_clients(private_key, is_testnet)
        asset_info, asset_id = get_asset_info(info_client, signal.symbol)

        strategy_position, is_buy, db_id, oid = get_strategy_position(
            info_client, user_address, signal.symbol, signal.strategy
        )

        sz_decimals = int(asset_info.get("szDecimals", 0))
        market_price = get_formatted_market_price(info_client, signal.symbol, is_buy, sz_decimals)

        # Build order
        # bulk_orders expects OrderRequest objects (not raw wires)
        order = {
            "coin": signal.symbol,
            "is_buy": not is_buy,
            "sz": float(strategy_position),
            "limit_px": float(market_price),
            "order_type": {"limit": {"tif": "Gtc"}},
            "reduce_only": True
        }

        # Execute using bulk_orders
        close_response = exchange_client.bulk_orders([order])

        # Parse the response dict
        try:
            statuses = close_response.get("response", {}).get("data", {}).get("statuses", [])
        except AttributeError:
            if isinstance(close_response, dict) and 'response' in close_response:
                statuses = close_response['response']['data']['statuses']
            else:
                statuses = close_response.get("status", [])

        if not statuses:
            raise AppError("Order response missing statuses", HTTP.BAD_REQUEST)

        order_status = statuses[0]
        if 'error' in order_status:
            raise AppError(f"Failed to close position: {order_status['error']}", HTTP.BAD_REQUEST)

        close_order_id = extract_order_id(order_status)
        pnl_data = get_order_pnl(info_client, user_address, str(oid))

        update_order(
            id=db_id,
            status='closed',
            pnl=pnl_data.get('netPnl'),
            stopLossOid=str(close_order_id)
        )

        send_notification(
            title="Order Closed",
            symbol=signal.symbol,
            is_buy=is_buy,
            price=signal.price,
            stop_loss=signal.stopLoss,
            position_value=signal.positionValue,
            pnl=pnl_data
        )

        return OrderResult(
            success=True,
            message=f"Closed {signal.symbol} position",
            orderId=str(oid),
            stopLossOrderId=str(close_order_id),
            dbOrderId=db_id
        )

    except Exception as error:
        return OrderResult(
            success=False,
            error=str(error)
        )

File: src/hyperLiquid_ts_deprecated/services/closeOrder.ts
import { WebhookPayload, OrderResult } from "../models/webhook";
import * as hl from "@nktkas/hyperliquid";
import { formatPrice } from "@nktkas/hyperliquid/utils";
import { updateOrder } from "../repositories/dynamo.repository";
import { AppError } from "../helpers/errorHandler";
import { HTTP } from "../constants/http";
import { sendNotification } from "../helpers/telegram";
import { getMarketPrice } from "../helpers/marketPrice.helpers";

import {
    getEnvConfig,
    createClients,
    getAssetInfo,
    getStrategyPosition,
    getOrderPnl,
    extractOrderId
} from "../helpers/hyperliquid.helpers";
/**
 * Cancel all open orders for a symbol (including stop loss)
 * @param exchangeClient - HyperLiquid exchange client
 * @param infoClient - HyperLiquid info client
 * @param userAddress - User's wallet address
 * @param symbol - Trading symbol
 * @param assetId - Asset ID from HyperLiquid meta data
 */
async function cancelOpenOrders(
    exchangeClient: hl.ExchangeClient,
    infoClient: hl.InfoClient,
    userAddress: string,
    symbol: string,
    assetId: number
): Promise<void> {
    const openOrders = await infoClient.openOrders({ user: userAddress as `0x${string}` });
    const ordersToCancel = openOrders.filter(order => order.coin === symbol);

    if (ordersToCancel.length > 0) {
        await exchangeClient.cancel({
            cancels: ordersToCancel.map(order => ({
                a: assetId,
                o: order.oid
            }))
        });
        console.log(`Cancelled ${ordersToCancel.length} open orders for ${symbol}`);
    }
}

/**
 * Close an open position by placing a limit order at near-market price
 * Cancels all open orders, calculates PnL, and sends notification
 * @param signal - Webhook payload containing EXIT signal
 * @param context - Optional Azure Function context for logging
 * @returns Order result with success status and message
 */
export async function closeOrder(
    signal: WebhookPayload,
    context?: any
): Promise<OrderResult> {
    try {
        console.log(`[closeOrder] Received signal:`, JSON.stringify(signal, null, 2));

        const { privateKey, userAddress, isTestnet } = getEnvConfig();
        const { exchangeClient, infoClient } = createClients(privateKey, isTestnet);
        const { assetInfo, assetId } = await getAssetInfo(infoClient, signal.symbol);

        // Get current position from HyperLiquid and Database
        const { strategyPosition, isBuy, id, oid } = await getStrategyPosition(infoClient, userAddress, signal.symbol, signal.strategy);

        // Get near-market price for quick execution
        const closePrice = await getMarketPrice(infoClient, signal.symbol, isBuy);
        const marketPrice = formatPrice(closePrice, assetInfo.szDecimals);

        const closeResponse = await exchangeClient.order({
            orders: [{
                a: assetId,
                b: !isBuy, // Opposite of current position
                p: marketPrice,
                s: strategyPosition,
                r: true, // Reduce-only to ensure it only closes position
                t: { limit: { tif: "Gtc" } } // Good-til-cancel to ensure full position closure
            }],
            grouping: "na"
        });

        const orderStatus = closeResponse.response.data.statuses[0];
        const closeOrderId = extractOrderId(orderStatus);
        if ('error' in orderStatus) {
            throw new AppError(`Failed to close position: ${orderStatus.error}`, HTTP.BAD_REQUEST);
        }

        const pnlData = await getOrderPnl(infoClient, userAddress, oid);
        await updateOrder(id, {
            status: 'closed',
            pnl: pnlData.netPnl,
            stopLossOid: closeOrderId.toString()
        });

        await sendNotification(
            "Order Closed",
            signal.symbol,
            isBuy,
            signal.price,
            signal.stopLoss,
            signal.positionValue
        );

        return {
            success: true,
            message: `Closed ${signal.symbol} position`,
            orderId: oid.toString(),
            stopLossOrderId: closeOrderId.toString(),
            dbOrderId: id
        };

    } catch (error) {
        return {
            success: false,
            error: error instanceof Error ? error.message : "Unknown error"
        };
    }
}